File: src/terminal/xterm-keys.lua
local function make_generator_node(impl)
    return function(buf, indent)
        for line in impl:gmatch("[^\n]+") do
            buf:write(indent, line, "\n")
        end
    end
end

local function set_key_and_goto(key, label)
    local template = "tmp = %s;\ngoto %s;\n"
    return make_generator_node(template:format(key, label))
end

local xterm_keys = {
    ["[A"] = "KEY_UP",
    ["[B"] = "KEY_DOWN",
    ["[C"] = "KEY_RIGHT",
    ["[D"] = "KEY_LEFT",
    ["[F"] = "KEY_END",
    ["[H"] = "KEY_HOME",
    ["[L"] = "KEY_INSERT",
    ["[Z"] = "MOD_SHIFT | '\\t'",

    ["OA"] = "KEY_UP",
    ["OB"] = "KEY_DOWN",
    ["OC"] = "KEY_RIGHT",
    ["OD"] = "KEY_LEFT",
    ["OF"] = "KEY_END",
    ["OH"] = "KEY_HOME",
    ["OI"] = "'\\t'",
    ["OM"] = "'\\r'",
    ["OP"] = "KEY_F1",
    ["OQ"] = "KEY_F2",
    ["OR"] = "KEY_F3",
    ["OS"] = "KEY_F4",
    ["OX"] = "'='",
    ["Oj"] = "'*'",
    ["Ok"] = "'+'",
    ["Ol"] = "','",
    ["Om"] = "'-'",
    ["Oo"] = "'/'",
    ["O "] = "' '",

    ["[1~"] = "KEY_HOME",
    ["[2~"] = "KEY_INSERT",
    ["[4~"] = "KEY_END",

    -- Generators for Ctrl/Meta/Shift key combinations
    ["[1;"] = make_generator_node("return parse_csi1(buf, length, i, k);"),
    ["[2;"] = set_key_and_goto("KEY_INSERT", "check_modifiers"),
    ["[3"] = set_key_and_goto("KEY_DELETE", "check_delim"),
    ["[5"] = set_key_and_goto("KEY_PAGE_UP", "check_delim"),
    ["[6"] = set_key_and_goto("KEY_PAGE_DOWN", "check_delim"),
    ["[15"] = set_key_and_goto("KEY_F5", "check_delim"),
    ["[17"] = set_key_and_goto("KEY_F6", "check_delim"),
    ["[18"] = set_key_and_goto("KEY_F7", "check_delim"),
    ["[19"] = set_key_and_goto("KEY_F8", "check_delim"),
    ["[20"] = set_key_and_goto("KEY_F9", "check_delim"),
    ["[21"] = set_key_and_goto("KEY_F10", "check_delim"),
    ["[23"] = set_key_and_goto("KEY_F11", "check_delim"),
    ["[24"] = set_key_and_goto("KEY_F12", "check_delim"),

    -- These 4 sequences are "deprecated" in xterm, even though they
    -- were used by default in older releases. They are still used by
    -- some other, xterm-like terminal emulators and can still be
    -- activated in xterm by setting the "oldXtermFKeys" resource.
    ["[11~"] = "KEY_F1",
    ["[12~"] = "KEY_F2",
    ["[13~"] = "KEY_F3",
    ["[14~"] = "KEY_F4",

    -- Linux console
    ["[[A"] = "KEY_F1",
    ["[[B"] = "KEY_F2",
    ["[[C"] = "KEY_F3",
    ["[[D"] = "KEY_F4",
    ["[[E"] = "KEY_F5",
}

-- Build a trie structure from tables, indexed by numeric byte values
local function make_trie(keys)
    local trie = {}
    for seq, key in pairs(keys) do
        local node = trie
        local n = #seq
        for i = 1, n do
            local char = assert(seq:sub(i, i))
            local byte = assert(char:byte())
            if i == n then
                -- Leaf node
                assert(node[byte] == nil, "Clashing sequence: " .. seq)
                node[byte] = key
                break
            elseif type(node[byte]) ~= "table" then
                assert(node[byte] == nil, "Clashing sequence: " .. seq)
                node[byte] = {}
            end
            node = node[byte]
        end
    end
    return trie
end

-- Create a caching lookup table of indent strings
local indents = setmetatable({[0] = "", [1] = "    "}, {
    __index = function(self, depth)
        assert(depth < 10)
        local indent = self[1]:rep(depth)
        self[depth] = indent
        return indent
    end
})

local function has_only_tilde_leaf(node)
    for i = ("\0"):byte(), ("}"):byte() do
        if node[i] then
            return false
        end
    end
    return type(node[("~"):byte()]) == "string"
end

-- Iterate over trie nodes and generate equivalent, nested C switch statements
local function write_trie(node, buffer)
    local buf = buffer or Buffer()
    local function serialize(node, depth)
        local indent = indents[depth]
        local node_type = type(node)
        if node_type == "table" then
            if has_only_tilde_leaf(node) then
                buf:write (
                    indent, "tmp = ", node[("~"):byte()], ";\n",
                    indent, "goto check_trailing_tilde;\n"
                )
                return
            end
            buf:write (
                -- Return -1 to indicate truncation, if buffer matches a
                -- prefix but isn't long enough to form a complete sequence
                indent, "if (i >= length) return -1;\n",
                indent, "switch(buf[i++]) {\n"
            )
            for i = ("\0"):byte(), ("~"):byte() do
                local v = node[i]
                if v then
                    buf:write(indent, "case '", string.char(i), "':\n")
                    serialize(v, depth + 1)
                end
            end
            buf:write(indent, "}\n", indent, "return 0;\n")
        elseif node_type == "function" then
            node(buf, indent)
        else
            assert(node_type == "string")
            buf:write (
                indent, "*k = ", node, ";\n",
                indent, "return i;\n"
            )
        end
    end
    serialize(node, 1)
end

local xterm_trie = make_trie(xterm_keys)
local output = assert(io.stdout)

output:write [[
// Escape sequence parser for xterm function keys.
// Copyright 2018 Craig Barnes.
// SPDX-License-Identifier: GPL-2.0-only
// Generated by: src/terminal/xterm-keys.lua
// See also: https://invisible-island.net/xterm/ctlseqs/ctlseqs.html

#include "xterm.h"

// These values are used in xterm escape sequences to indicate
// modifier+key combinations.
// See also: user_caps(5)
static KeyCode mod_enum_to_mod_mask(char mod_enum)
{
    switch (mod_enum) {
    case '2': return MOD_SHIFT;
    case '3': return MOD_META;
    case '4': return MOD_SHIFT | MOD_META;
    case '5': return MOD_CTRL;
    case '6': return MOD_SHIFT | MOD_CTRL;
    case '7': return MOD_META | MOD_CTRL;
    case '8': return MOD_SHIFT | MOD_META | MOD_CTRL;
    }
    return 0;
}

static ssize_t parse_csi1(const char *buf, size_t length, size_t i, KeyCode *k)
{
    if (i >= length) {
        return -1;
    }
    const KeyCode mods = mod_enum_to_mod_mask(buf[i++]);
    if (mods == 0) {
        return 0;
    } else if (i >= length) {
        return -1;
    }
    switch(buf[i++]) {
    case 'A': *k = mods | KEY_UP; return i;
    case 'B': *k = mods | KEY_DOWN; return i;
    case 'C': *k = mods | KEY_RIGHT; return i;
    case 'D': *k = mods | KEY_LEFT; return i;
    case 'F': *k = mods | KEY_END; return i;
    case 'H': *k = mods | KEY_HOME; return i;
    case 'P': *k = mods | KEY_F1; return i;
    case 'Q': *k = mods | KEY_F2; return i;
    case 'R': *k = mods | KEY_F3; return i;
    case 'S': *k = mods | KEY_F4; return i;
    }
    return 0;
}

ssize_t xterm_parse_key(const char *buf, size_t length, KeyCode *k)
{
    if (length == 0 || buf[0] != '\033') {
        return 0;
    }
    KeyCode tmp;
    size_t i = 1;
]]

write_trie(xterm_trie, output)

output:write [[
check_delim:
    if (i >= length) return -1;
    switch(buf[i++]) {
    case ';':
        goto check_modifiers;
    case '~':
        *k = tmp;
        return i;
    }
    return 0;
check_modifiers:
    if (i >= length) {
        return -1;
    }
    const KeyCode mods = mod_enum_to_mod_mask(buf[i++]);
    if (mods == 0) {
        return 0;
    }
    tmp |= mods;
check_trailing_tilde:
    if (i >= length) {
        return -1;
    } else if (buf[i++] == '~') {
        *k = tmp;
        return i;
    }
    return 0;
}
]]

File: src/terminal/xterm-keys.c
// Escape sequence parser for xterm function keys.
// Copyright 2018-2019 Craig Barnes.
// SPDX-License-Identifier: GPL-2.0-only
// See also: https://invisible-island.net/xterm/ctlseqs/ctlseqs.html

#include <stdint.h>
#include "xterm.h"
#include "../util/ascii.h"
#include "../util/macros.h"
#include "../util/unicode.h"

static const KeyCode modifiers[] = {
    [2] = MOD_SHIFT,
    [3] = MOD_META,
    [4] = MOD_SHIFT | MOD_META,
    [5] = MOD_CTRL,
    [6] = MOD_SHIFT | MOD_CTRL,
    [7] = MOD_META | MOD_CTRL,
    [8] = MOD_SHIFT | MOD_META | MOD_CTRL,
};

static const KeyCode special_keys[] = {
    [1] = KEY_HOME,
    [2] = KEY_INSERT,
    [3] = KEY_DELETE,
    [4] = KEY_END,
    [5] = KEY_PAGE_UP,
    [6] = KEY_PAGE_DOWN,
    [7] = KEY_HOME,
    [8] = KEY_END,
    [11] = KEY_F1,
    [12] = KEY_F2,
    [13] = KEY_F3,
    [14] = KEY_F4,
    [15] = KEY_F5,
    [17] = KEY_F6,
    [18] = KEY_F7,
    [19] = KEY_F8,
    [20] = KEY_F9,
    [21] = KEY_F10,
    [23] = KEY_F11,
    [24] = KEY_F12,
};

static KeyCode decode_modifiers(uint32_t n)
{
    return (n >= ARRAY_COUNT(modifiers)) ? 0 : modifiers[n];
}

static KeyCode decode_special_key(uint32_t n)
{
    return (n >= ARRAY_COUNT(special_keys)) ? 0 : special_keys[n];
}

// Fix quirky key codes sent when "modifyOtherKeys" is enabled
static KeyCode normalize_modified_other_key(KeyCode mods, KeyCode key)
{
    if (key > 0x20 && key < 0x80) {
        // The Shift modifier is never appropriate with the
        // printable ASCII range, since pressing Shift causes
        // the base key itself to change (i.e. "r" becomes "R',
        // "." becomes ">", etc.)
        mods &= ~MOD_SHIFT;
        if (mods & MOD_CTRL) {
            // The Ctrl modifier should always cause letters to
            // be uppercase -- this assumption is too ingrained
            // and causes too much breakage if not enforced
            key = ascii_toupper(key);
        }
    }
    return mods | keycode_normalize(key);
}

static ssize_t parse_ss3(const char *buf, size_t length, size_t i, KeyCode *k)
{
    if (i >= length) {
        return -1;
    }
    const char ch = buf[i++];
    switch (ch) {
    case 'A': // Up
    case 'B': // Down
    case 'C': // Right
    case 'D': // Left
    case 'F': // End
    case 'H': // Home
        *k = KEY_UP + (ch - 'A');
        return i;
    case 'M':
        *k = KEY_ENTER;
        return i;
    case 'P': // F1
    case 'Q': // F2
    case 'R': // F3
    case 'S': // F4
        *k = KEY_F1 + (ch - 'P');
        return i;
    case 'X':
        *k = '=';
        return i;
    case ' ':
        *k = ch;
        return i;
    case 'a': // Ctrl+Up (rxvt)
    case 'b': // Ctrl+Down (rxvt)
    case 'c': // Ctrl+Right (rxvt)
    case 'd': // Ctrl+Left (rxvt)
        *k = MOD_CTRL | (KEY_UP + (ch - 'a'));
        return i;
    case 'j': case 'k': case 'l':
    case 'm': case 'n': case 'o':
    case 'p': case 'q': case 'r':
    case 's': case 't': case 'u':
    case 'v': case 'w': case 'x':
    case 'y': case 'I':
        *k = ch - 64;
        return i;
    }
    return 0;
}

static ssize_t parse_csi_num(const char *buf, size_t len, size_t i, KeyCode *k)
{
    uint32_t params[4] = {0, 0, 0, 0};
    size_t nparams = 0;
    uint8_t final_byte = 0;

    uint32_t num = 0;
    size_t digits = 0;
    while (i < len) {
        const char ch = buf[i++];
        switch (ch) {
        case '0': case '1': case '2': case '3': case '4':
        case '5': case '6': case '7': case '8': case '9':
            num = (num * 10) + (ch - '0');
            if (num > UNICODE_MAX_VALID_CODEPOINT) {
                return 0;
            }
            digits++;
            continue;
        case ';':
            params[nparams++] = num;
            if (nparams > 2) {
                return 0;
            }
            num = 0;
            digits = 0;
            continue;
        case 'A': case 'B': case 'C': case 'D': case 'F':
        case 'H': case 'P': case 'Q': case 'R': case 'S':
        case 'u': case '~':
            final_byte = ch;
            if (digits > 0) {
                params[nparams++] = num;
            }
            goto exit_loop;
        }
        return 0;
    }
exit_loop:

    if (final_byte == 0) {
        return (i >= len) ? -1 : 0;
    }

    KeyCode mods = 0;
    KeyCode key;

    switch (nparams) {
    case 3:
        if (params[0] == 27 && final_byte == '~') {
            mods = decode_modifiers(params[1]);
            if (mods == 0) {
                return 0;
            }
            *k = normalize_modified_other_key(mods, params[2]);
            return i;
        }
        return 0;
    case 2:
        mods = decode_modifiers(params[1]);
        if (mods == 0) {
            return 0;
        }
        switch (final_byte) {
        case '~':
            goto check_first_param_is_special_key;
        case 'u':
            *k = normalize_modified_other_key(mods, params[0]);
            return i;
        case 'A': // Up
        case 'B': // Down
        case 'C': // Right
        case 'D': // Left
        case 'F': // End
        case 'H': // Home
            key = KEY_UP + (final_byte - 'A');
            goto check_first_param_is_1;
        case 'P': // F1
        case 'Q': // F2
        case 'R': // F3
        case 'S': // F4
            key = KEY_F1 + (final_byte - 'P');
            goto check_first_param_is_1;
        }
        return 0;
    case 1:
        if (final_byte == '~') {
            goto check_first_param_is_special_key;
        }
        return 0;
    }
    return 0;

check_first_param_is_special_key:
    key = decode_special_key(params[0]);
    if (key == 0) {
        return 0;
    }
    goto set_k_and_return_i;
check_first_param_is_1:
    if (params[0] != 1) {
        return 0;
    }
set_k_and_return_i:
    *k = mods | key;
    return i;
}

static ssize_t parse_csi(const char *buf, size_t length, size_t i, KeyCode *k)
{
    if (i >= length) {
        return -1;
    }
    char ch = buf[i++];
    switch (ch) {
    case 'A': // Up
    case 'B': // Down
    case 'C': // Right
    case 'D': // Left
    case 'F': // End
    case 'H': // Home
        *k = KEY_UP + (ch - 'A');
        return i;
    case 'a': // Shift+Up (rxvt)
    case 'b': // Shift+Down (rxvt)
    case 'c': // Shift+Right (rxvt)
    case 'd': // Shift+Left (rxvt)
        *k = MOD_SHIFT | (KEY_UP + (ch - 'a'));
        return i;
    case 'L':
        *k = KEY_INSERT;
        return i;
    case 'Z':
        *k = MOD_SHIFT | '\t';
        return i;
    case '0': case '1': case '2': case '3': case '4':
    case '5': case '6': case '7': case '8': case '9':
        return parse_csi_num(buf, length, i - 1, k);
    case '[':
        if (i >= length) {
            return -1;
        }
        switch (ch = buf[i++]) {
        // Linux console keys
        case 'A': // F1
        case 'B': // F2
        case 'C': // F3
        case 'D': // F4
        case 'E': // F5
            *k = KEY_F1 + (ch - 'A');
            return i;
        }
        return 0;
    }
    return 0;
}

ssize_t xterm_parse_key(const char *buf, size_t length, KeyCode *k)
{
    if (length == 0 || buf[0] != '\033') {
        return 0;
    } else if (length == 1) {
        return -1;
    }
    switch (buf[1]) {
    case 'O': return parse_ss3(buf, length, 2, k);
    case '[': return parse_csi(buf, length, 2, k);
    }
    return 0;
}
